### PR TITLE
Multi datasource support

### PR DESCRIPTION
**Pull request by sumukhswamy**
### Description
mds support for neo
cache changes 
### Issues Resolved
https://github.com/opensearch-project/dashboards-query-workbench/issues/266 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: common/utils/fetch_datasources.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+          const connector = item.connector;
+          const name = item.name;
+
+          if (connector === 'S3GLUE') {
+            if (!connectorGroups[connector]) {
+              connectorGroups[connector] = [];
+            }
+
+            connectorGroups[connector].push(name);
+            if (name === urlDataSource) {
```
Comment (BionIT) [inline common/utils/fetch_datasources.ts:31]: what is urlDataSource? Can we give a better name?

File: public/components/QueryLanguageSwitch/ClusterTabs.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { EuiButtonGroup } from '@elastic/eui';
+import React from 'react';
+// @ts-ignore
```
Comment (BionIT) [inline public/components/QueryLanguageSwitch/ClusterTabs.tsx:8]: what are we ignoring?

File: public/components/QueryLanguageSwitch/ClusterTabs.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    id: 'Data Connections',
+    label: 'Data Connections',
+  },
+];
+
+class ClusterTabs extends React.Component<SwitchProps> {
+  constructor(props: SwitchProps) {
+    super(props);
+    this.state = {
+      cluster: 'Indexes',
```
Comment (BionIT) [inline public/components/QueryLanguageSwitch/ClusterTabs.tsx:31]: do we use this anywhere?

File: public/components/SQLPage/SQLPage.tsx [hunk 4]
```diff
@@ -39,6 +39,7 @@ interface SQLPageProps {
   asyncLoading: boolean;
   openAccelerationFlyout: boolean;
   setIsAccelerationFlyoutOpened: (value: boolean) => void;
+  dataSourceClientId: string;
```
Comment (BionIT) [inline public/components/SQLPage/SQLPage.tsx:42]: is this dataSourceId?

File: public/components/SQLPage/sql_catalog_tree/s3_tree.tsx [hunk 5]
```diff
@@ -218,19 +226,19 @@ export const S3Tree = ({ dataSource, updateSQLQueries, refreshTree }: S3TreeProp
...
-    startDatabasesLoading(dataSource);
+    startDatabasesLoading({dataSourceName: dataSource, dataSourceId: dataSourceId});
   };
 
   const onLoadS3Tree = () => {
     setIsTreeLoading({ status: true, message: '' });
-    const dsCache = catalogCacheRefs.CatalogCacheManager!.getOrCreateDataSource(dataSource);
-
+    const dsCache = catalogCacheRefs.CatalogCacheManager!.getOrCreateDataSource(dataSource, dataSourceId);
+    console.log(dsCache)
```
Comment (BionIT) [inline public/components/SQLPage/sql_catalog_tree/s3_tree.tsx:235]: do we want to clean this up?

**Comment by BionIT** [comment]
Can we add test details with screenshots or recording?

**Review by joshuali925 — COMMENTED** [review]
didn't finish, will review later

File: common/utils/fetch_datasources.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+import { CoreStart } from "../../../../src/core/public";
+
+export const fetchDataSources = (http: CoreStart['http'], dataSourceId: string, urlDataSource: string, onSuccess, onError) => {
```
[the remark this answers is not in the document]
Comment (sumukhswamy) [reply common/utils/fetch_datasources.ts:9]: this one here actually refers to the flint datasource id

File: public/components/SQLPage/SQLPage.tsx (hunk 4 above)
Comment (sumukhswamy) [reply public/components/SQLPage/SQLPage.tsx:42]: to distinguish between flint datasource-ids and mds ids I have named it like this

File: public/components/QueryLanguageSwitch/ClusterTabs.tsx (hunk 3 above)
Comment (sumukhswamy) [reply public/components/QueryLanguageSwitch/ClusterTabs.tsx:31]: The main reason I have a state here is to distinguish between indexes tab and data connections, which I check in
https://github.com/opensearch-project/dashboards-query-workbench/pull/311/files#diff-3ff08cae50890804a95e00ef9f0f2a560c5ef12a4075e322612f7008bcdaa978R23

**Comment by sumukhswamy** [comment]
Uploading Screen Recording 2024-04-25 at 4.50.21 PM.mov…

**Comment by sumukhswamy** [comment]
https://github.com/opensearch-project/dashboards-query-workbench/assets/107440906/61da59b4-55a9-4aed-a05c-0ea932ca2c98

File: common/utils/async_query_helpers.ts [hunk 7]
```diff
@@ -41,6 +42,9 @@ export const executeAsyncQuery = (
           ...query,
           sessionId: getAsyncSessionId(currentDataSource) ?? undefined,
         }),
+        query: {
+          dataSourceMDSId: dataSourceMDSId,
```
Comment (BionIT) [inline common/utils/async_query_helpers.ts:46]: as suggested, we can use shorthand here
```suggestion
          dataSourceMDSId,
```

File: server/services/QueryService.ts [hunk 8]
```diff
@@ -62,17 +98,29 @@ export default class QueryService {
...
           resp: _.isEqual(responseFormat, 'json') ? JSON.stringify(queryResponse) : queryResponse,
         },
       };
     } catch (err) {
-      console.log(err);
+      this.logger.info('error describeQueryGetInternalSync');
+      this.logger.info(err);
+      this.logger.info(request.query);
+
+      console.log(err, request.query);
```
Comment (BionIT) [inline server/services/QueryService.ts:123]: can we clean up the console print?

File: server/services/QueryService.ts [hunk 9]
```diff
@@ -39,19 +60,34 @@ export default class QueryService {
...
         data: {
           ok: true,
           resp: _.isEqual(responseFormat, 'json') ? JSON.stringify(queryResponse) : queryResponse,
         },
       };
     } catch (err) {
-      console.log(err);
+      this.logger.info('error describeQueryJobIdInternal');
+      this.logger.info(err);
+      this.logger.info(request.query);
```
Comment (BionIT) [inline server/services/QueryService.ts:88]: do we intentionally add those logs?

File: opensearch_dashboards.json [hunk 10]
```diff
@@ -1,9 +1,9 @@
 {
   "id": "queryWorkbenchDashboards",
-  "version": "3.0.0.0",
-  "opensearchDashboardsVersion": "3.0.0",
+  "version": "2.12.0.0",
+  "opensearchDashboardsVersion": "2.12.0",
```
Comment (BionIT) [inline opensearch_dashboards.json:4]: are we intentionally changing to 2.12 here?

File: public/components/Main/main.tsx [hunk 11]
```diff
@@ -542,12 +590,16 @@ export class Main extends React.Component<MainProps, MainState> {
...
+        query = { dataSourceMDSId: this.state.selectedMDSDataConnectionId };
+      }
       const endpoint =
         '/api/sql_console/' + (_.isEqual(language, 'SQL') ? 'translatesql' : 'translateppl');
       const translationPromise = Promise.all(
-        queries.map((query: string) =>
+        queries.map((eachQuery: string) =>
           this.httpClient
-            .post(endpoint, { body: JSON.stringify({ query }) })
+            .post(endpoint, { body: JSON.stringify({ query: eachQuery }), query })
```
Comment (joshuali925) [inline public/components/Main/main.tsx:602]: this is sometimes `payload`, sometimes `eachQuery`, might be better to keep consistent

File: public/components/QueryLanguageSwitch/ClusterTabs.tsx [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  asyncLoading: boolean;
+}
+
+const toggleButtons = [
+  {
+    id: 'Indexes',
+    label: 'Indexes',
+  },
+  {
+    id: 'Data source Connections',
```
Comment (joshuali925) [inline public/components/QueryLanguageSwitch/ClusterTabs.tsx:21]: np. i think usually id should be something like `data-source-connections`, separate from display name. if display name changes, id doesn't need to be changed

File: server/services/QueryService.ts [hunk 13]
```diff
@@ -7,28 +7,49 @@
...
           resp: _.isEqual(responseFormat, 'json') ? JSON.stringify(queryResponse) : queryResponse,
         },
       };
     } catch (err) {
-      console.log(err);
+      this.logger.info('error describeQueryPostInternal');
+      this.logger.info(err);
+      this.logger.info(request.query);
+
+      console.log(err, request.query);
```
Comment (zhyuanqi) [inline server/services/QueryService.ts:52]: Multiple logger and console.log here

File: common/utils/fetch_datasources.ts [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    http.get(`/api/get_datasources/${dataSourceMDSId}`)
+      .then((res) => {
+        const data = res.data.resp;
+        const connectorGroups = {};
+
+        data.forEach((item) => {
+          const connector = item.connector;
+          const name = item.name;
```
Comment (zhyuanqi) [inline common/utils/fetch_datasources.ts:23]: const { connector, name } = item; We can deconstructing the item here

File: common/utils/fetch_datasources.ts [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+            if (!connectorGroups[connector]) {
+              connectorGroups[connector] = [];
+            }
+
+            connectorGroups[connector].push(name);
+            if (name === urlDataSource) {
+              urlSourceFound = true;
+            }
+          }
+        });
```
Comment (zhyuanqi) [inline common/utils/fetch_datasources.ts:35]: We can simplify the logic here
```
if (connector === 'S3GLUE') {
        connectorGroups[connector] = connectorGroups[connector] || []; 
        connectorGroups[connector].push(name);

        if (name === urlDataSource) {
          urlSourceFound = true;
        }
      }

File: public/components/Main/main.tsx [hunk 16]
```diff
@@ -809,11 +877,37 @@ export class Main extends React.Component<MainProps, MainState> {
...
   };
 
+  onChangeCluster = (id: string) => {
+    this.setState({
+      cluster: id,
+      queryResultsTable: [],
+    });
+    if (id === 'Indexes') {
+      this.setState({
+        selectedDatasource: [{ label: 'OpenSearch', key: '' }],
```
[the remark this answers is not in the document]
Comment (sumukhswamy) [reply public/components/Main/main.tsx:887]: selected datasource is mainly for flint

File: public/application.tsx [hunk 17]
```diff
@@ -7,21 +7,28 @@
...
 export const renderApp = (
-  { notifications, http, chrome }: CoreStart,
-  { navigation }: AppPluginStartDependencies,
-  { appBasePath, element }: AppMountParameters
+  { notifications, http, chrome, savedObjects }: CoreStart,
+  { navigation, dataSource }: AppPluginStartDependencies,
+  { appBasePath, element, setHeaderActionMenu }: AppMountParameters,
+  dataSourceManagement: DataSourceManagementPluginSetup
 ) => {
+  console.log(dataSource)
```
Comment (joshuali925) [inline public/application.tsx:20]: remove this?

File: public/components/SQLPage/sql_catalog_tree/s3_tree_helpers.tsx [hunk 18]
```diff
@@ -25,6 +25,8 @@ import { CachedDataSourceStatus, TreeItem, TreeItemType } from '../../../../comm
 import { useToast } from '../../../../common/utils/toast_helper';
 import { catalogCacheRefs } from '../../../framework/catalog_cache_refs';
 
+let pageStateLanguage: string;
+// const [language, setPageLanguage] = useState('SQL');
```
Comment (joshuali925) [inline public/components/SQLPage/sql_catalog_tree/s3_tree_helpers.tsx:29]: remove this comment?

File: public/components/SQLPage/sql_catalog_tree/s3_tree_helpers.tsx [hunk 19]
```diff
@@ -36,12 +38,30 @@ export const handleQuery = (
   updateSQLQueries(`select * from \`${dataSource}\`.\`${database}\`.\`${tableName}\` limit 10`);
 };
 
+export const pageLanguage = (
+  language: string
+) => {
+  pageStateLanguage = language
+}
```
Comment (joshuali925) [inline public/components/SQLPage/sql_catalog_tree/s3_tree_helpers.tsx:45]: what is this for? also usually functions should start with verbs

File: public/components/SQLPage/sql_catalog_tree/s3_tree_helpers.tsx [hunk 20]
```diff
@@ -54,7 +74,13 @@ export const createLabel = (
               {node.type === TREE_ITEM_TABLE_NAME_DEFAULT_NAME && !node.isLoading && (
                 <EuiIcon
                   type="editorCodeBlock"
-                  onClick={(e) => handleQuery(e, dataSource, database, node.name, updateSQLQueries)}
+                  onClick={(e) => {
+                    if (pageStateLanguage === 'SQL') {
```
Comment (joshuali925) [inline public/components/SQLPage/sql_catalog_tree/s3_tree_helpers.tsx:78]: prefer to use state or prop

File: server/services/QueryService.ts (hunk 13 above)
Comment (joshuali925) [inline server/services/QueryService.ts:52]: remove console.log since they are already logged

File: server/services/QueryService.ts [hunk 21]
```diff
@@ -39,19 +60,32 @@ export default class QueryService {
...
       return {
         data: {
           ok: true,
           resp: _.isEqual(responseFormat, 'json') ? JSON.stringify(queryResponse) : queryResponse,
         },
       };
     } catch (err) {
-      console.log(err);
+      this.logger.info(err);
+      this.logger.info(request.query);
```
Comment (joshuali925) [inline server/services/QueryService.ts:87]: similar question as above

File: public/components/SQLPage/sql_catalog_tree/catalog_tree.tsx [hunk 22]
```diff
@@ -12,23 +12,43 @@ interface CatalogTreeProps {
...
         />
+      ):(
+        <EuiFlexItem grow={false}>
+          <EuiEmptyPrompt
+            icon={<EuiIcon type="database" size="m" />}
+            iconColor="subdued"
+            titleSize="xs"
+            body={<p>Select a Data Source Connection to fetch Databases</p>}
+          />
+        </EuiFlexItem>
```
Comment (ps48) [inline public/components/SQLPage/sql_catalog_tree/catalog_tree.tsx:51]: Nested ternary operators are difficult to understand, maybe can be split in variables like done inside the tree code.

**Comment by ps48** [comment]
@sumukhswamy Can you please create an issue for fast follow on the linter issues? We can fix these after this PR is merged.

**Review by ps48 — APPROVED** [review]

**Pull request merged**